staging/src/k8s.io/externaljwt/pkg/plugin/keycache.go

```diff
@@ -0,0 +1,203 @@
+			continue
+		}
+
+		p.broadcastUpdate()
+	}
+}
+
+func (p *KeyCache) AddListener(listener Listener) {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	p.listeners = append(p.listeners, listener)
+}
+
+func (p *KeyCache) GetCacheAgeMaxSeconds() int {
+	// consumer shall support the keys for no longer than 10 mins before refreshing.
+	return p.refreshIntervalSec
+}
+
+// GetPublicKeys returns the public key corresponding to requested keyID.
+// Getter is expected to return All keys for keyID ""
+func (p *KeyCache) GetPublicKeys(keyID string) []PublicKey {
+	ctx := context.Background()
+
+	pubKeys, ok := p.findKeyForKeyID(keyID)
+	if ok {
+		return pubKeys
+	}
+
+	// If we didn't find it, trigger a sync.
+	if err := p.syncKeys(ctx); err != nil {
+		klog.ErrorS(err, "Error while syncing keys")
+	}
+
+	pubKeys, ok = p.findKeyForKeyID(keyID)
+	if ok {
+		return pubKeys
+	}
+
+	// If we still didn't find it, then it's an unknown keyID.
+	klog.ErrorS(nil, "Key id not found after refresh", "keyID", keyID)
+	return []PublicKey{}
+}
+
+func (p *KeyCache) findKeyForKeyID(keyID string) ([]PublicKey, bool) {
+	p.publicKeysLock.RLock()
+	defer p.publicKeysLock.RUnlock()
+
+	if len(p.publicKeys.Keys) == 0 {
+		klog.ErrorS(nil, "No keys currently in cache.  Initial fill has not completed")
+		return nil, false
+	}
+
+	if keyID == "" {
+		return p.publicKeys.Keys, true
+	}
+
+	for _, key := range p.publicKeys.Keys {
+		if key.KeyID == keyID {
+			return []PublicKey{key}, true
+		}
+	}
+
+	return nil, false
+}
+
+// sync supported external keys.
+// completely re-writes the set of supported keys.
+func (p *KeyCache) syncKeys(ctx context.Context) error {
+	_, err, _ := p.syncGroup.Do("", func() (any, error) {
+		newPublicKeys, refreshHint, err := p.plugin.GetTokenVerificationKeys(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("while fetching token verification keys: %w", err)
+		}
+
+		p.publicKeysLock.Lock()
+		defer p.publicKeysLock.Unlock()
+
+		p.publicKeys = newPublicKeys
+		p.updateRefreshInterval(refreshHint)
+
+		return newPublicKeys, nil
+	})
+	if err != nil {
+		return fmt.Errorf("in singleflight: %w", err)
+	}
+
+	return nil
+}
+
+func (p *KeyCache) metrics() (float64, []string) {
+	p.publicKeysLock.RLock()
+	defer p.publicKeysLock.RUnlock()
+
+	if p.publicKeys == nil {
+		return -1.0, nil
+	}
+
+	keyIDs := make([]string, len(p.publicKeys.Keys))
+	for i := 0; i < len(p.publicKeys.Keys); i++ {
+		keyIDs[i] = p.publicKeys.Keys[i].KeyID
+	}
+	return time.Since(p.publicKeys.DataTimestamp).Seconds(), keyIDs
+}
+
+func (p *KeyCache) broadcastUpdate() {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	for _, l := range p.listeners {
+		l.Enqueue()
+	}
+}
+
+func (p *KeyCache) updateRefreshInterval(refreshHint int) {
+	if refreshHint > 0 && p.refreshIntervalSec != refreshHint {
+		p.refreshIntervalSec = refreshHint
+		if p.ticker != nil {
+			p.ticker.Reset(time.Second * time.Duration(p.refreshIntervalSec))
```

Comment:
I think there might be some confusion about what this field actually means.  I thought we had said it was 'in this many seconds, you should refresh next' which is not the same as an interval for refreshing (i.e. I think we want a timer, not a ticker).

Reply:
Agree it's confusing as written currently.

A timer runs / expires / fires once, which matches what we expect in the happy path.

Need to decide what interval we recheck on when the periodic refresh fails.

Reply:
Redid it